[Contact] ContactTitle, Form, 개인정보 활용동의 컴포넌트 + emailJS 연결 초안

ingbinsee opened the pull request:
## 🔍️ 이 PR을 통해 해결하려는 문제가 무엇인가요?

> 어떤 기능을 구현한건지, 이슈 대응이라면 어떤 이슈인지 PR이 열리게 된 계기와 목적을 Reviewer 들이 쉽게 이해할 수 있도록 적어 주세요
> 일감 백로그 링크나 다이어그램, 피그마를 첨부해도 좋아요

- Contact 페이지의 기능 및 레이아웃(header, main, nav, footer 등 공통 section 제외)을 구현하였습니다.

## ✨ 이 PR에서 핵심적으로 변경된 사항은 무엇일까요?

> 문제를 해결하면서 주요하게 변경된 사항들을 적어 주세요

- 분기점(모바일, 태블릿, 데스크탑)에 따른 반응형 레이아웃 구현
- Form 컴포넌트, 개인정보 활용동의 컴포넌트 제작
- emailJS 서비스 연결(설정된 이메일로 작성된 내용 발송)

https://github.com/ReOpenSG/ReOpenSG/assets/140426866/760a4971-d064-4610-b794-236d64e19544

## 🔖 핵심 변경 사항 외에 추가적으로 변경된 부분이 있나요?

> 없으면 "없음" 이라고 기재해 주세요

- 없음

## 🙏 Reviewer 분들이 이런 부분을 신경써서 봐 주시면 좋겠어요

> 개발 과정에서 다른 분들의 의견은 어떠한지 궁금했거나 크로스 체크가 필요하다고 느껴진 코드가 있다면 남겨주세요

- 없음

## 🩺 이 PR에서 테스트 혹은 검증이 필요한 부분이 있을까요?

> 테스트가 필요한 항목이나 테스트 코드가 추가되었다면 함께 적어주세요

- 없음

### 📌 PR 진행 시 이러한 점들을 참고해 주세요

- Reviewer 분들은 코드 리뷰 시 좋은 코드의 방향을 제시하되, 코드 수정을 강제하지 말아 주세요.
- Reviewer 분들은 좋은 코드를 발견한 경우, 칭찬과 격려를 아끼지 말아 주세요.
- Review는 특수한 케이스가 아니면 Reviewer로 지정된 시점 기준으로 3일 이내에 진행해 주세요.
- Comment 작성 시 Prefix로 P1, P2, P3 를 적어 주시면 Assignee가 보다 명확하게 Comment에 대해 대응할 수 있어요
  - P1 : 꼭 반영해 주세요 (Request Changes) - 이슈가 발생하거나 취약점이 발견되는 케이스 등
  - P2 : 반영을 적극적으로 고려해 주시면 좋을 것 같아요 (Comment)
  - P3 : 이런 방법도 있을 것 같아요~ 등의 사소한 의견입니다 (Chore)

#

---

## 📝 Assignee를 위한 CheckList

- [ ] To-Do Item

ingbinsee committed:
✨ [Feat] - Form 컴포넌트 작성

Contact 페이지의 문의유형 및 문의내용 작성 폼을 추가하였습니다.
+ 문의유형 클릭 시 단일 선택이 가능하도록 하였습니다.
+ react-toastify를 활용하여 전송 성공과 실패 메시지를 보여주도록 하였습니다.

! 문의유형 아이콘은 임의로 넣어두었습니다.
! 반응형을 고려한 스타일링이 완료되지 않았습니다.

Related to: #13

ingbinsee committed:
🤝 [Merge] : 원격 저장소와 로컬 저장소 동기화

ingbinsee committed:
⚙️ [Chore] : 공통 환경 세팅 변경

eslint 일부 옵션을 off하였습니다.

ingbinsee committed:
✨ [Feat] - Form 컴포넌트 완료

Form 컴포넌트를 반응형으로 구현하였습니다.
+ 분기점(모바일, 태블릿, 데스크탑)에 따른 변화를 반영하였습니다.
+ module.css로 스타일을 분리하였습니다.
+ FormSection.jsx 파일을 Contact.jsx로 통합하였습니다.
+ 아이콘의 크기를 조절하였습니다.

! 문의유형 아이콘은 임의로 넣어두었습니다.

Related to: #13

ingbinsee committed:
✨ [Feat] : 개인정보 활용동의 컴포넌트 완료

Contact 페이지에 개인정보 활용동의를 추가하였습니다.
+ 필수 동의 미체크 시 react-toastify를 활용하여 에러 메시지를 보여주도록 하였습니다.
+ 분기점(모바일, 태블릿, 데스크탑)에 따른 변화를 반영하였습니다.

Related to: #21

ingbinsee committed:
✨ [Feat] : Contact 페이지 Title 컴포넌트 완료

Title 컴포넌트를 구현하였습니다.
+ 분기점(모바일, 태블릿, 데스크탑)에 따른 변화를 반영하였습니다.

Related to: #28

ingbinsee committed:
♻️ [Refactor] : 컴포넌트 분리 및 리뷰 사항 반영

+ 하나의 컴포넌트 파일에 작성된 코드를 컴포넌트로 별도 분리하였습니다.
+ <hr />로 되어 있던 마크업을 border 스타일링으로 변경하였습니다.
+ 배열을 활용하여 data를 구성하고 이를 map 함수를 활용하는 방법으로 문의유형 카테고리를 구성하였습니다.
+ 시안 변경에 따라 문의 유형 부분을 justify-content: space-between에서 gap: 16px(모바일), 36px(데스크탑, 태블릿)으로 변경하였습니다.

! 문의유형 클릭 시 아이콘 색상이 변경되어야 하나 해결하지 못했습니다.

Related to: #7

ingbinsee committed:
⚙️ [Chore] : 공통 환경 세팅 변경

eslint 일부 옵션을 off하였습니다.

ingbinsee committed:
🤝 [Merge] : 원격 저장소와 로컬 저장소 동기화

ingbinsee committed:
📂 [Rename] : Button 컴포넌트명 변경

Button 컴포넌트명을 ContactButton으로 변경하였습니다.

ingbinsee committed:
🏛️ [Style] : tailwind.config 파일 변경에 따른 스타일링 수정

tailwind.config 파일에서 spacing 값이 변경됨에 따라 스타일링을 수정하였습니다.

ingbinsee committed:
📂 [Rename] : contact 폴더명 변경

contact 폴더명을 Contact로 변경하였습니다.

ingbinsee committed:
♻️ [Refactor] : 리뷰 사항 반영

+ 하나의 컴포넌트 파일에 작성하였던 Title 컴포넌트를 별도 컴포넌트로 분리하였습니다.
+ svg 컴포넌트의 분기점별 크기와 높이를 40px(모바일), 80px(데스크탑, 태블릿)로 지정하였습니다.
+ prop-types를 불러오는 방식을 string에서 PropTypes.string로 변경하였습니다.
+ defaultProps를 없애고 isRequired를 사용하여 기본값 지정 대신 필수 요소로 지정하는 방식으로 변경하였습니다.
+ input 요소에 required 속성을 부여하여 필수 기입 요소로 지정하였습니다.
+ 구조 분해 할당으로 Depth가 깊어지는 것을 방지하였습니다.

ingbinsee committed:
⚙️ [Chore] : 공통 환경 세팅 변경

+ ToastContainer를 RootLayout에서 1회 사용하여 공통으로 활용할 수 있도록 하였습니다.
+ tailwind.config.js에서 spacing의 변경된 크기를 반영하였습니다.

ingbinsee committed:
🤝 [Merge] : 원격 저장소와 로컬 저장소 동기화

ingbinsee committed:
♻️ [Refactor] : 리뷰 사항 반영
+ 문의유형 선택 시 이미지와 텍스트의 색상이 변경되도록 하였습니다.

Related to: #13

ingbinsee committed:
➖ [Remove] : 불필요한 파일 삭제

svg 이미지를 컴포넌트로 작업함에 따라 불필요한 이미지 파일을 삭제하였습니다.

ingbinsee committed:
⚙️ [Chore] : 공통 환경 세팅 변경

eslint 일부 옵션을 off하였습니다.

ingbinsee committed:
✨ [Feat] : 문의유형 아이콘 변경을 위한 조건 처리

문의유형별 텍스트에 따라 아이콘이 변경될 수 있도록 조건 처리를 하였습니다.

Related to: #13

ingbinsee committed:
🌈 [Design] : module.css로 CSS 정리

tailwindcss로 작업된 디자인 코드를 module.css로 변경하였습니다.

ingbinsee committed:
🤝 [Merge] : 원격 저장소와 로컬 저장소 동기화

ryudg commented on src/components/Contact/Agreement.jsx:
```diff
@@ -0,0 +1,65 @@
+import React, { useState } from 'react';
+import { toast } from 'react-toastify';
+import styles from '@/components/contact/Contact.module.css';
```
**P1**: `'@/components/contact/Contact.module.css'` 실제 경로명 `'@/components/Contact/Contact.module.css'`으로 변경해주세요

ingbinsee replied:
변경하였습니다.

ryudg commented on src/components/Contact/Agreement.jsx:
```diff
@@ -0,0 +1,65 @@
+            id="agreement"
+            required
+            onClick={handleChangeCheck}
+            className={`peer ${styles.checkboxInput}`}
```
**P2** : 
기존 `onClick` prop보다 `onChange` prop으로 변경이 적절해 보입니다. (input이 **클릭**하는 이벤트 -> input의 값이 **변경**되는 이벤트)
이후, input 요소의 `checked` 속성에 `agreementStatus` 상태를 연결해 UI와 state간의 일치를 시켜주는게 좋을것 같습니다.
```jsx
<input
    // 생략...
    checked={agreementStatus}
    onChange={handleChangeCheck}
/>
```
이 경우 스크린 리더 및 보조 기술에서 `checked` 속성을 사용해 체크박스의 상태를 정확히 전달할 수 있습니다.
즉, `onClick` 이벤트만 사용하는 경우 스크린 리더와 같은 보조 기술은 실제로 체크박스의 상태를 감지하지 못할 수 있습니다.

ingbinsee replied:
디테일한 부분을 생각하지 못했습니다. 감사합니다.
반영하였습니다.

ryudg commented on src/components/Contact/Agreement.jsx:
```diff
@@ -0,0 +1,65 @@
+
+  return (
+    <>
+      <fieldset>
```
**P3**: 기존 중복되는 스타일 적용 코드를 개선하면 좋을것 같습니다.

```jsx
// src/components/Contact/Agreement.jsx

// ...생략
    <fieldset className={styles.agreementText}>
    <h3 className={styles.agreementTitle}>개인정보 활용동의</h3>
    <p className={`my-open-xl tablet:my-open-2xl desktop:my-open-2xl`}>
        OPENSG는 개인정보보호법 및 관련 법령에 의거 고객님들의 의견 접수와 관련한 성실한 답변을
        드리기 위해 필요한 최소한의 개인정보를 아래와 같이 수집, 이용하고자 합니다.
    </p>
    <ol className={styles.agreementOl}>
        <li>
        필수정보의 수집 목적 : 온라인 상담실을 통한 정보주체의 의견접수 및 문의에 대한 답변/안내
        </li>
        <li>
        필수 수집 항목 : 이름, 이메일 주소
        </li>
        <li>
        보유 및 이용기간 : 상담 신청 후 1년
        </li>
        <li>
        ※ 고객님께서는 본 동의를 거부하실 수 있으나 거부 시 필수정보의 수집 목적에 해당하는
        사항의 처리가 제한될 수 있습니다.
        </li>
    </ol>
    <div className={styles.checkboxWrapper}>
        <input
        type="checkbox"
        id="agreement"
        required
        onClick={handleChangeCheck}
        className={`peer ${styles.checkboxInput}`}
        />
        <label htmlFor="agreement" className={`pl-open-lg`}>
        (필수) 위 개인정보 수집 및 이용에 동의합니다.
        </label>
    </div>
    </fieldset>
// 생략...
```

```css
/* src/Contact/Contact.module.css */

/* 생략... */
    .agreementOl {
        @apply rounded-[20px] border -border--open-gray-300 px-open-lg tablet:px-open-xl desktop:px-open-xl py-open-xl tablet:py-open-2xl desktop:py-open-2xl;

        & li {
            @apply list-inside list-decimal;
        }
    }
/* 생략... */
```

ingbinsee replied:
스타일을 중복하여 사용하고 있음을 인지하지 못했네요. 감사합니다.
반영하였습니다.

ryudg commented on src/components/Contact/ContactTitleInner.jsx:
```diff
@@ -0,0 +1,25 @@
+import React from 'react';
+import PropTypes from 'prop-types';
+import styles from '@/components/contact/Contact.module.css';
```
**P1**: `@/components/contact/Contact.module.css` 실제 경로명 `@/components/Contact/Contact.module.css`으로 변경해주세요

ingbinsee replied:
변경하였습니다.

ryudg commented on src/components/Contact/ContactTitleSection.jsx:
```diff
@@ -0,0 +1,17 @@
+import React from 'react';
+import styles from '@/components/contact/Contact.module.css';
```
**P1**: `@/components/contact/Contact.module.css` 실제 경로명 `@/components/Contact/Contact.module.css`으로 변경해주세요

ingbinsee replied:
변경하였습니다.

ryudg commented on src/components/Contact/Content.jsx:
```diff
@@ -0,0 +1,35 @@
+import React from 'react';
+import { v4 as uuidv4 } from 'uuid';
+import FillingInput from './FillingInput';
+import styles from '@/components/contact/Contact.module.css';
```
**P1**: `@/components/contact/Contact.module.css` 실제 경로명 `@/components/Contact/Contact.module.css`으로 변경해주세요

ingbinsee replied:
변경하였습니다.

ryudg commented on src/components/Contact/InquiryInput.jsx:
```diff
@@ -0,0 +1,40 @@
+import PropTypes from 'prop-types';
+import React from 'react';
+import styles from '@/components/contact/Contact.module.css';
```
**P1**: `@/components/contact/Contact.module.css` 실제 경로명 `@/components/Contact/Contact.module.css`으로 변경해주세요

ingbinsee replied:
변경하였습니다.

ryudg commented on src/components/Contact/InquiryType.jsx:
```diff
@@ -0,0 +1,27 @@
+import React from 'react';
+import InquiryInput from './InquiryInput';
+import { data } from '@/data/contactData';
+import styles from '@/components/contact/Contact.module.css';
```
**P1**: `@/components/contact/Contact.module.css` 실제 경로명 `@/components/Contact/Contact.module.css`으로 변경해주세요

ingbinsee replied:
변경하였습니다.

ryudg commented on src/components/Contact/icons/ProductIcon.jsx:
```diff
@@ -0,0 +1,22 @@
+import React from 'react';
+import styles from '@/components/contact/Contact.module.css';
```
**P1**: `@/components/contact/Contact.module.css` 실제 경로명 `@/components/Contact/Contact.module.css`으로 변경해주세요

ingbinsee replied:
변경하였습니다.

ryudg commented on src/components/Contact/PersonalInfo.jsx:
```diff
@@ -0,0 +1,24 @@
+import React from 'react';
+import styles from '@/components/contact/Contact.module.css';
```
**P1**: `@/components/contact/Contact.module.css` 실제 경로명 `@/components/Contact/Contact.module.css`으로 변경해주세요

ingbinsee replied:
변경하였습니다.

ryudg commented on src/views/Contact.jsx:
```diff
@@ -1,7 +1,38 @@
+import React, { useRef } from 'react';
+import emailjs from '@emailjs/browser';
+import { toast } from 'react-toastify';
+import PersonalInfo from '@/components/contact/PersonalInfo';
+import InquiryType from '@/components/contact/InquiryType';
+import Content from '@/components/contact/Content';
+import Agreement from '@/components/contact/Agreement';
```
**P1**: `@/components/contact/Contact.module.css` 실제 경로명 `@/components/Contact/Contact.module.css`으로 변경해주세요

ingbinsee replied:
변경하였습니다.

ingbinsee committed:
🌈 [Design] : max-width 설정

제약 없이 늘어나는 콘텐츠 크기를 조정하기 위해 max-width(1320px)를 설정하였습니다.

ingbinsee committed:
🐛 [Fix]  : 폴더명 변경에 따른 import 수정

폴더명을 contact에서 Contact로 변경함에 따라 module.css를 불러오는 경로를 수정하였습니다.

ryudg commented on src/components/Contact/Agreement.jsx:
```diff
@@ -0,0 +1,65 @@
+          </label>
+        </div>
+      </fieldset>
+      <ContactButton
```
**P3**: `ContactButton` 컴포넌트는 `form` 요소의 제출을 위한 버튼 컴포넌트도 확인 됩니다.
이 경우 "개인정보 활용동의"(`Agreement`) 컴포넌트 내부보다 `form` 요소 자체(`src/views/Contact`)에 구성되어있는게 가독성이 좋아보입니다.

ingbinsee replied:
ContactButton 컴포넌트를 &lt;form&gt;이 있는 ```src/views/Contact.jsx```로 이동하였습니다.

ryudg reviewed (changes requested):
요청 사항 확인하고 변경해주세요

kelvin-chihyun commented on src/views/Contact.jsx:
```diff
@@ -1,7 +1,38 @@
+
+  const sendEmail = (e) => {
+    e.preventDefault();
+    emailjs.sendForm('opensg', 'opensg', form.current, '6NuaSgBDx54b9MicI').then(
```
P1: API 관련 부분은 env에 변수로 보관해주세요!

ingbinsee replied:
API 관련 내용 env.local에 설정 후 불러오는 방식으로 수정하였습니다.

kelvin-chihyun commented on src/components/Contact/Contact.module.css:
P2: Nesting 가능한 자식 태그는 부모 태그 아래에 두면, 가독성이 조금 더 좋아질 듯 합니다!

ingbinsee replied:
Nesting 가능한 자식 태그를 부모 태그 아래로 두었습니다.

kelvin-chihyun reviewed (changes requested):
P1: API 키는 보안 이슈가 있으니, 수정해주시길 바라겠습니다!

ryudg reviewed (approved):
수정 사항 모두 확인했습니다.

ingbinsee committed:
♻️ [Refactor] : 리뷰 사항 반영

+ 개인정보 활용동의 props를 onClick에서 onChange로 변경하였습니다.
+ input 요소의 checked 속성에 agreementStatus 상태를 연결하였습니다.
+ ContactButton 컴포넌트를 <form>이 있는 src/views/Contact.jsx로 이동하였습니다.
+ API 관련 내용은 보안을 위해 env.local에 설정 후 불러오는 방식으로 수정하였습니다.
+ 중복된 스타일 코드 사용을 개선하였습니다.
+ Nesting 가능한 자식 태그를 부모 태그 아래로 두었습니다.

Related to: #42

ingbinsee committed:
Merge branch 'develop' into contact

kelvin-chihyun reviewed (approved):
확인했습니다!
